### ISSUE TITLE
dlopen on Chrome

### ISSUE THREAD

**Issue by mdboom** [kripken/emscripten#6633]
As is well known in these parts, Chrome requires that WebAssembly is compiled asynchronously.

I can make that work for the MAIN_MODULE, using `WebAssembly.compileStreaming`.  However, for the `SIDE_MODULE`, I'm not sure how to make this work.  I've experimented with rewriting `loadWebAssemblyModule` to use the asynchronous APIs, but then I have the issue that it's called from within `dlopen`, which needs to at least "feel" synchronous on the C side of things.

Do you have any advice or pointers on how to make this work?  Are there any other examples of asynchronous work in JS that we fake to look synchronous on the C side?

**Comment by kripken** [comment]
One option: is the code calling it dependent on it being synchronous? What I mean is, we could add an async variant of C's `dlopen`, if you can use it.

Another option is we could add a JS API to pre-compile the libraries. Something like this: user input indicates that a specific side module is needed. You call an async JS API to fetch it from the remote server, and a new JS API to do an async compile in preparation for running in the main module. You then run the synchronous C code, and the dlopen code internally finds the already-compiled module and just uses it synchronously.

If none of those are practical, the "big hammer" we have here is the emterpreter. It can make sync C code async. Emterpreted code is very slow, but only it and code above it on the stack needs to be emterpreted, so it's possible the speed-sensitive areas can remain at full speed. See https://github.com/kripken/emscripten/wiki/Emterpreter#emterpreter-async-run-synchronous-code

**Comment by mdboom** [comment]
I've decided to proceed with option B, and I have something that mostly works.  I'm instantiating the WebAssembly when the filesystem data is loaded, and then actually doing the dlopen linking synchronously later.

However, I'm running into the following issue when instantiating (`WebAssembly.instantiate`) the modules.  It happens after a certain number have been instantiated, and doesn't seem to be related to particular modules.  On Chrome, I get:

```
Uncaught (in promise) LinkError: WebAssembly Instantiation: table initializer is out of bounds
```
and on Firefox
```
LinkError: elem segment does not fit in table
```

I'm having trouble finding information on what these messages mean, but I'm guessing some sort of function lookup table is overflowing from loading so much code at once...?

One thing that's counter to my theory is that in Firefox, it goes on to successfully load the failing .wasm modules synchronously, so it's not simply that all of these modules together overflow some table.

**Comment by mdboom** [comment]
If this is meaningful: It's always the same 4 wasm modules that fail to load asynchronously.
Looking at `loadWebAssemblyModule` which parses a `memorySize` and `tableSize` from the file, I see that these failing ones all have a table size of `1024` or higher.  The ones that consistently pass have a table size of `512` or lower.

**Comment by mdboom** [comment]
Maybe I'm running into this issue because `BINARYEN_ASYNC_COMPILATION` is disabled due to `EMULATED_FUNCTION_POINTERS`...

**Comment by mdboom** [comment]
Aha!  I've found the solution.  (Apologies for spamming my thought processes above).

`loadWebAssemblyModule` is not 'threadsafe', in that it uses  global state to set things up for the module (`env['memoryBase']` and `env['tableBase']`, for example).  When you call `loadWebAssemblyModule` asynchronously, this setup and the actual instantiation of the module (by `WebAssembly.instantiate`) may happen out of order, and one module will get the settings intended for another.  The quick fix seems to be compile the wasm modules "asynchronously" (to address Chrome's requirement) but in series.

https://files.gitter.im/iodide-project/iodide/NL5p/image.png

**Comment by kripken** [comment]
Interesting!

There's probably a way to let them compile asynchronously in parallel, but might not be worth the effort. In sequence as you said sounds good.

**Comment by mdboom** [comment]
I think this is fixed by the merging of #6663
